eureka2-ext/eureka2-overrides-aws/src/main/java/com/netflix/eureka2/server/service/overrides/S3OutOfServiceOverridesSource.java

```diff
@@ -0,0 +1,141 @@
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectResult;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.netflix.eureka2.registry.instance.Delta;
+import com.netflix.eureka2.registry.instance.InstanceInfo;
+import com.netflix.eureka2.registry.instance.InstanceInfoField;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An optimized s3 backed overrides source for OUT_OF_SERVICE status overrides only.
+ * In most cases overrides are only needed for this specific case, and given this constraint,
+ * we can actually make use of s3 object names for storage.
+ *
+ * The file format will be of the form: {config.getPrefix()}/overrideId
+ * where if the file is present, then the OOS override exist
+ *
+ * @author David Liu
+ */
+@Singleton
+public class S3OutOfServiceOverridesSource implements LoadingOverridesRegistry.ExternalOverridesSource {
+
+    private static final Logger logger = LoggerFactory.getLogger(S3OutOfServiceOverridesSource.class);
+
+    private final S3OverridesConfig config;
+    private final AmazonS3Client s3Client;
+
+    @Inject
+    public S3OutOfServiceOverridesSource(S3OverridesConfig config) {
+        this(config, new AmazonS3Client());
+    }
+
+    public S3OutOfServiceOverridesSource(S3OverridesConfig config, AmazonS3Client s3Client) {
+        this.config = config;
+        this.s3Client = s3Client;
+    }
+
+    @Override
+    public void set(Overrides overrides) throws Exception {
+        if (overrides.getDeltas().size() != 1) {
+            throw new UnsupportedOperationException("Not handling multiple overrides");
+        }
+
+        Delta<?> delta = overrides.getDeltas().iterator().next();
+        if (delta.getField().getValueType() != InstanceInfo.Status.class) {
+            throw new UnsupportedOperationException("Not handling non-status overrides");
+        }
+
+        InstanceInfo.Status status = (InstanceInfo.Status) delta.getValue();
+        if (status != InstanceInfo.Status.OUT_OF_SERVICE) {
+            throw new UnsupportedOperationException("Not handling non OUT_OR_SERVICE overrides");
+        }
+
+        InputStream inputStream = new ByteArrayInputStream(new byte[0]);
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(0);
+
+        String name = toS3Name(config.getPrefix(), overrides.getId());
+
+        PutObjectResult result = s3Client.putObject(
+                config.getBucketName(),
+                name,
+                inputStream,
+                metadata
+        );
+
+        logger.debug("S3 put result: {}", result);
```

Comment:
If operation fails, will this result in an exception or error code will be send in the result? In the latter case we should have error logging added.